### PR TITLE
add optional doughnut parameter to contract calls

### LINKED ISSUE #33: Implement hook for contract to contract calls
Implement the hook from #32  for contract to contract calls.

- [ ] Add doughnut field to the contract ExecutionContext
https://github.com/plugblockchain/plug-blockchain/blob/b93e3585eb9bb9f1648a0d155e67832a8aa7f30c/srml/contracts/src/exec.rs#L263

- [ ] Expose execution context doughnut via new doughnut() method on the Ext trait
https://github.com/plugblockchain/plug-blockchain/blob/b93e3585eb9bb9f1648a0d155e67832a8aa7f30c/srml/contracts/src/exec.rs#L93

- [ ] Add contract-to-contract hook using `Ext::doughnut()`
https://github.com/plugblockchain/plug-blockchain/blob/b93e3585eb9bb9f1648a0d155e67832a8aa7f30c/srml/contracts/src/wasm/runtime.rs#L385

### PR DESCRIPTION
**Pull request by cowboy-bebug**
Changes:
- [x] ExecutionContext has optional doughnut field
    - added optional doughnut param to methods where needed
    - for tests and node/runtime, None is used as doughnut for now
- [x]  Expose execution context doughnut via new doughnut() method on the Ext trait
- [x] Add contract-to-contract hook using Ext::doughnut()

Closes #33

### CODE REVIEW

File: frame/contracts/src/wasm/mod.rs [hunk 1]
```diff
@@ -903,6 +909,49 @@ mod tests {
...
+		)
+	)
+
+	;; Load a storage value into the scratch buf.
+	(func (export "call")
+		;; assert $ext_scratch_size == 8
+		(call $assert
+			(i32.eq
+				(call $ext_doughnut)
+				(i32.const 0)
```
Comment (jordy25519) [inline frame/contracts/src/wasm/mod.rs]: This is constant `0`, comment above says `1`

File: frame/contracts/src/wasm/mod.rs [hunk 2]
```diff
@@ -903,6 +909,49 @@ mod tests {
 		).unwrap();
 	}
 
+	/// calls `ext_doughnut`, loads the doughnut from the scratch buffer and
```
Comment (jordy25519) [inline frame/contracts/src/wasm/mod.rs]: This test is only checking the return status code: it should assert the contents of the buffer aswell, like with this one: https://github.com/plugblockchain/plug-blockchain/blob/37c0fcfb34923ce5692932b76fe97b5f7cb68580/frame/contracts/src/wasm/mod.rs#L880-L895

File: frame/contracts/src/wasm/runtime.rs [hunk 3]
```diff
@@ -560,6 +560,18 @@ define_env!(Env, <E: Ext>,
 		Ok(())
 	},
 
+	// Stores the doughnut of the current contract into the scratch buffer.
+	// Returns 0 if doughnut is present, otherwise 1.
+	ext_doughnut(ctx) -> u32 => {
```
Comment (jordy25519) [inline frame/contracts/src/wasm/runtime.rs]: We don't need this function to be implemented yet, if at all

File: frame/contracts/src/exec.rs [hunk 4]
```diff
@@ -284,6 +288,7 @@ pub struct ExecutionContext<'a, T: Trait + 'a, V, L> {
 	pub timestamp: MomentOf<T>,
 	pub block_number: T::BlockNumber,
 	pub origin: T::AccountId,
+	pub doughnut: Option<T::Doughnut>,
```
Comment (jordy25519) [inline frame/contracts/src/exec.rs]: We should make this by ref if possible `Option<&T::Doughnut>` and avoid all the clones

**Review by jordy25519 — CHANGES_REQUESTED** [review]
PR looks good in general
- Don't need `ext_doughnut` call to be implemented for this PR but left feedback anyway ;)
- Pass the doughnut by ref if possible (may require change to @alexsednz s work)

File: frame/contracts/src/wasm/runtime.rs (hunk 3 above)
Comment (cowboy-bebug) [reply frame/contracts/src/wasm/runtime.rs]: Thanks! I'll remove it and the related test 😄

File: frame/contracts/src/exec.rs [hunk 5]
```diff
@@ -284,6 +288,7 @@ pub struct ExecutionContext<'a, T: Trait + 'a, V, L> {
 	pub timestamp: MomentOf<T>,
 	pub block_number: T::BlockNumber,
 	pub origin: T::AccountId,
+	pub doughnut: Option<&'a T::Doughnut>,
```
Comment (aliXsed) [inline frame/contracts/src/exec.rs:291]: Wouldn't it be better that we pass the doughnut in through RawOrigin::Signed(issuer, doughnut) as it is for RawOrigin::Delegated for the sake of consistency?

Comment (cowboy-bebug) [reply frame/contracts/src/exec.rs:291]: Yeah I wasn't sure which doughnut to use so just used one from `system::Trait`

Comment (aliXsed) [reply frame/contracts/src/exec.rs:291]: Yup, if we do so then we wouldn't have two doughnuts passed in directly and indirectly through RawOrigin.

Comment (cowboy-bebug) [reply frame/contracts/src/exec.rs:291]: I see what u mean but they are the same thing as `system::Trait` is defined as:
```rust
pub trait Trait: 'static + Eq + Clone {
    /// The aggregated `Origin` type used by dispatchable calls.
    type Origin:
        Into<Result<RawOrigin<Self::AccountId, Self::Doughnut>, Self::Origin>> + 
        From<RawOrigin<Self::AccountId, Self::Doughnut>> +
    ...
}
```
`Self::Doughnut` inside Origin is the `T::Doughnut`

Comment (aliXsed) [reply frame/contracts/src/exec.rs:291]: That's why I suggest we should only have one of them. If they could be different then there could be scenarios where we would need to get either of these doughnuts separately and that could justify passing in two references. Does it make sense?

Comment (cowboy-bebug) [reply frame/contracts/src/exec.rs:291]: It's more clear and consistent to me that we are using `system::Trait::Doughnut` type, similar to how RawOrigin and DelegatedDispatchVerifier are also using `system::Trait::Doughnut` type. And we are not passing anything just yet, it's just declaring what the type should be for the doughnut field.

File: frame/contracts/src/wasm/runtime.rs [hunk 6]
```diff
@@ -387,6 +388,15 @@ define_env!(Env, <E: Ext>,
 		let value: BalanceOf<<E as Ext>::T> =
 			read_sandbox_memory_as(ctx, value_ptr, value_len)?;
 
+		if let Some(doughnut) = ctx.ext.doughnut() {
+			type DelegatedDispatchVerifierOf<E> = <<E as Ext>::T as system::Trait>::DelegatedDispatchVerifier;
+			DelegatedDispatchVerifierOf::<E>::verify_contract_to_contract_call(
+				&callee,
```
Comment (jordy25519) [inline frame/contracts/src/wasm/runtime.rs]: This one should be the `origin` address

File: frame/contracts/src/wasm/runtime.rs [hunk 7]
```diff
@@ -387,6 +388,15 @@ define_env!(Env, <E: Ext>,
 		let value: BalanceOf<<E as Ext>::T> =
 			read_sandbox_memory_as(ctx, value_ptr, value_len)?;
 
+		if let Some(doughnut) = ctx.ext.doughnut() {
+			type DelegatedDispatchVerifierOf<E> = <<E as Ext>::T as system::Trait>::DelegatedDispatchVerifier;
```
Comment (jordy25519) [inline frame/contracts/src/wasm/runtime.rs]: Can move the type alias to the top of the file

**Review by hoani — APPROVED** [review]
LGTM

**Review by jordy25519 — CHANGES_REQUESTED** [review]
Looks good! Should add a couple integration tests to check `verify_contract_to_contract_call` works correctly. forgot to request this for `runtime_to_contract` in #47

**Review by jordy25519 — APPROVED** [review]

**Comment by jordy25519** [comment]
LGTM 🎉 🎉 🎉

File: frame/contracts/src/lib.rs [hunk 8]
```diff
@@ -572,12 +574,13 @@ decl_module! {
 			dest: <T::Lookup as StaticLookup>::Source,
 			#[compact] value: BalanceOf<T>,
 			#[compact] gas_limit: Gas,
-			data: Vec<u8>
+			data: Vec<u8>,
+			doughnut: Option<T::Doughnut>,
```
Comment (jordy25519) [inline frame/contracts/src/lib.rs:578]: so `doughnut` should be extracted from the `origin`. the additional parameter here and in `instantiate` should be removed

Comment (cowboy-bebug) [reply frame/contracts/src/lib.rs:578]: Fixed by https://github.com/plugblockchain/plug-blockchain/pull/49

**Pull request merged**